File: partiql-plan/src/main/java/org/partiql/plan/rel/RelUnion.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical union abstract base class.
 */
public abstract class RelUnion extends RelBase {

    /**
     * @return new {@link RelUnion} instance
     */
    @NotNull
    public static RelUnion create(@NotNull Rel left, @NotNull Rel right, boolean all) {
        return new Impl(left, right, all);
    }

    /**
     * Whether this {@link RelUnion} set quantifier is ALL else DISTINCT.
     * @return true if ALL else DISTINCT.
     */
    public abstract boolean isAll();

    /**
     * Returns the left relation operator to a {@link RelUnion} relation operator (operator 0).
     * @return left rel (operand 0)
     */
    @NotNull
    public abstract Rel getLeft();

    /**
     * Returns the right relation operator to a {@link RelUnion} relation operator (operator 1).
     * @return right rel (operand 1)
     */
    @NotNull
    public abstract Rel getRight();

    @NotNull
    @Override
    protected final RelType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getLeft());
        Operand c1 = Operand.single(getRight());
        return List.of(c0, c1);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitUnion(this, ctx);
    }

    @NotNull
    public abstract RelUnion copy(@NotNull Rel left, @NotNull Rel right);

    @NotNull
    public abstract RelUnion copy(@NotNull Rel left, @NotNull Rel right, boolean all);

    private static class Impl extends RelUnion {

        private final Rel left;
        private final Rel right;
        private final boolean all;

        private Impl(Rel left, Rel right, boolean all) {
            this.left = left;
            this.right = right;
            this.all = all;
        }

        @Override
        public boolean isAll() {
            return all;
        }

        @NotNull
        @Override
        public Rel getLeft() {
            return left;
        }

        @NotNull
        @Override
        public Rel getRight() {
            return right;
        }

        @NotNull
        @Override
        public RelUnion copy(@NotNull Rel left, @NotNull Rel right) {
            return new Impl(left, right, all);
        }

        @NotNull
        @Override
        public RelUnion copy(@NotNull Rel left, @NotNull Rel right, boolean all) {
            return new Impl(left, right, all);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelUnion.kt
package org.partiql.plan.v1.rel

/**
 * Logical `UNION [ALL|DISTINCT]` operator for set (or multiset) union.
 */
public interface RelUnion : Rel {

    public fun isAll(): Boolean

    public fun getLeft(): Rel

    public fun getRight(): Rel

    override fun getInputs(): List<Rel> = listOf(getLeft(), getRight())

    override fun isOrdered(): Boolean = false

    public override fun <R, C> accept(visitor: RelVisitor<R, C>, ctx: C): R = visitor.visitUnion(this, ctx)

    /**
     * Default [RelUnion] implementation meant for extension.
     */
    public abstract class Base(left: Rel, right: Rel, all: Boolean) : RelUnion {

        /**
         * If neither ALL nor DISTINCT is specified, then DISTINCT is implicit (all = false).
         */
        public constructor(left: Rel, right: Rel) : this(left, right, false)

        // DO NOT USE FINAL
        private var _isAll = false
        private var _left = left
        private var _right = right
        private var _inputs: List<Rel>? = null

        override fun isAll(): Boolean = _isAll

        override fun getLeft(): Rel = _left

        override fun getRight(): Rel = _right

        override fun getInputs(): List<Rel> {
            if (_inputs == null) {
                _inputs = listOf(_left, _right)
            }
            return _inputs!!
        }

        override fun isOrdered(): Boolean = false

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is RelUnion) return false
            if (_isAll != other.isAll()) return false
            if (_left != other.getLeft()) return false
            if (_right != other.getRight()) return false
            return true
        }

        override fun hashCode(): Int {
            var result = _isAll.hashCode()
            result = 31 * result + _left.hashCode()
            result = 31 * result + _right.hashCode()
            return result
        }
    }
}
